src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs

```diff
@@ -0,0 +1,120 @@
+using Microsoft.ServiceHub.Framework.Services;
+using NuGet.Packaging;
+using NuGet.ProjectManagement;
+using NuGet.VisualStudio;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public sealed class NuGetProjectManagerService : INuGetProjectManagerService
+    {
+        private readonly ServiceActivationOptions _options;
+        private readonly IServiceBroker _serviceBroker;
+        private readonly AuthorizationServiceClient _authorizationServiceClient;
+
+        public NuGetProjectManagerService(ServiceActivationOptions options, IServiceBroker sb, AuthorizationServiceClient ac)
+        {
+            _options = options;
+            _serviceBroker = sb;
+            _authorizationServiceClient = ac;
+        }
+
+        public async ValueTask<IReadOnlyCollection<IProjectContextInfo>> GetProjectsAsync(CancellationToken cancellationToken)
+        {
+            var solutionManager = await ServiceLocator.GetInstanceAsync<IVsSolutionManager>();
+            Assumes.NotNull(solutionManager);
+
+            NuGetProject[] projects = (await solutionManager.GetNuGetProjectsAsync()).ToArray();
+            var projectContexts = new List<IProjectContextInfo>(projects.Length);
+
+            foreach (NuGetProject nugetProject in projects)
+            {
+                var projectContext = await ProjectContextInfo.CreateAsync(nugetProject, cancellationToken);
+                projectContexts.Add(projectContext);
+            }
+
+            return projectContexts;
+        }
+
+        public async ValueTask<IProjectContextInfo> GetProjectAsync(string projectGuid, CancellationToken cancellationToken)
+        {
+            var solutionManager = await ServiceLocator.GetInstanceAsync<IVsSolutionManager>();
+            Assumes.NotNull(solutionManager);
+
+            NuGetProject project = await GetNuGetProjectMatchingProjectGuidAsync(projectGuid);
+            return await ProjectContextInfo.CreateAsync(project, cancellationToken);
+        }
+
+        public async ValueTask<IReadOnlyCollection<PackageReference>> GetInstalledPackagesAsync(IReadOnlyCollection<string> projectGuids, CancellationToken cancellationToken)
+        {
+            var solutionManager = await ServiceLocator.GetInstanceAsync<IVsSolutionManager>();
+            Assumes.NotNull(solutionManager);
+
+            NuGetProject[]? projects = (await solutionManager.GetNuGetProjectsAsync())
+                .Where(p => projectGuids.Contains(p.GetMetadata<string>(NuGetProjectMetadataKeys.ProjectId)))
+                .ToArray();
+
+            // Read package references from all projects.
+            IEnumerable<Task<IEnumerable<PackageReference>>>? tasks = projects.Select(project => project.GetInstalledPackagesAsync(cancellationToken));
+            IEnumerable<PackageReference>[]? packageReferences = await Task.WhenAll(tasks);
+
+            return packageReferences.SelectMany(e => e).ToArray();
```

Comment:
Since we only need a subset of the information (just package identity and target framework), I wonder if we should be deduping before returning.

Imagine if you have 100 projects and each references the same package.  We'll return 100 references here.  Clients can/may dedupe, but it's noisy.

Reply:
I agree, but this is basically how it was before and we can optimize it in a later PR if we need to